### ISSUE TITLE
Normal Maps with large height gradients are not getting applied correctly

### ISSUE THREAD

**Issue by fuzzypurplepixel** [godotengine/godot#44285]
**Godot version:**
Godot 3.2.3 and 3.2.4-beta3

**OS/device including version:**
Windows 10, Nvidia GTX 960 GLES2 GLES3

**Issue description:**
The normal file in question is labeled Asteroid_DeepNormal.png as that has some height gradients. I also have an Asteroid_SoftNormal.png file that I made with softer gradients. The issue is I want some large height gradients for my normal maps to give an effect of deeper indents in my asteroids. I could see this have a use for other game objects as well.

The issue exists for GLES3 and GLES2 for both Sprite, Sprite3D, and MeshInstance. Parts of the sprite get shaded incorrectly when using normal maps that have large ranges in values. The issue does diminish when using the 3D methods and setting the normal scale to 0.5 or less. I have confirmed my normal map files are valid to work by testing on the Unity game engine.

Note about Godot Master branch: this issue is only for GLES2 and GLES3 when using a MeshInstance/Sprite3D as Sprite2D does not have a normal map option when making the issue using Godot Master branch with Volkan. I also can't seem to test the GLES2 on the master branch at this moment.

**Steps to reproduce:**
3DScene and 2DScene files are provided and should be runnable. It does seem that if the normal scale is set to 0.5 most of the issue is no longer visible.

**Minimal reproduction project:**
[TestNormals.zip](https://github.com/godotengine/godot/files/5676896/TestNormals.zip)

**Comment by clayjohn** [comment]
How are you generating the normal maps? Also are they HDR? it looks like the deep_normal texture might be HDR.

Also note it could be a precision related issue. The normal maps looks significantly better when they are uncompressed.

**Comment by fuzzypurplepixel** [comment]
> 
> 
> How are you generating the normal maps? Also are they HDR? it looks like the deep_normal texture might be HDR.
> 
> Also note it could be a precision related issue. The normal maps looks significantly better when they are uncompressed.

The normal maps are generated with tweaking in the Laigter tool. If you know how to use the tool this is a copy of my presets
[Presets.txt](https://github.com/godotengine/godot/files/5677337/Presets.txt)

I also have made a new version of the normals. I do run my images through a compressor before using them. This one I got right out of Laigter. This issue was still occurring before I was using the compression tool.
![Asteroid5D_Deep_uncompressed](https://user-images.githubusercontent.com/74986870/101869806-52f1d380-3b4e-11eb-8df8-31d9d465af55.png)

Another note. The images I provided are for testing with Godot only(don't redistribute them for any other purpose). I have designed them and plan on using them in my own commercial product.

**Comment by fuzzypurplepixel** [comment]
![Asteroid5D_n](https://user-images.githubusercontent.com/74986870/101870970-ab29d500-3b50-11eb-99f1-85a6a116814c.png)
Forgot to turn on the Invert y axis

**Comment by lawnjelly** [comment]
Sounds similar to #43309? (not that it got fully investigated)

**Comment by lawnjelly** [comment]
I made a test doing normal mapping type operation manually (just using a dot product of a light vector and the texture normal), and there were no problems with the texture. So it does suggest either the compression or the math in the shader.

![Screenshot from 2020-12-11 09-23-21](https://user-images.githubusercontent.com/21999379/101885890-9f5f0280-3b92-11eb-89ae-0ef712d168a7.png)

[normal_map_manual.zip](https://github.com/godotengine/godot/files/5678246/normal_map_manual.zip)

**Comment by lawnjelly** [comment]
Looks like it is the texture compression method. Same as before, except this time the shader loads the normal compressed as RG8, and uses the same math to derive the z of the normal as in the standard shaders. Result - horrible dithering.

![Screenshot from 2020-12-11 09-34-56](https://user-images.githubusercontent.com/21999379/101887068-48f2c380-3b94-11eb-8896-edc2fe5a52ab.png)

**Comment by lawnjelly** [comment]
This can be fixed in the shader. The correct version should be:
```
	normal.z = sqrt(1.0 - clamp(dot(normal.xy, normal.xy), 0, 1));
```

**Comment by fuzzypurplepixel** [comment]
![image](https://user-images.githubusercontent.com/74986870/101933289-fe7c4180-3ba9-11eb-9b9c-df35dba9ce5d.png)
This image is how it is rendered in Unity. Unity does have some more complex lighting options for 2D like a spotlight so it does have a higher contrast from the light source

**Comment by lawnjelly** [comment]
Interestingly, a lot of the dithering nastiness due to compression disappears when adding a bodge factor to the minimum. I'm not exactly sure why.. when x and y are 0, the tangent normal is pointing to the bottom left (or some corner) and the upward component is zero. Perhaps there's a large variation in normal direction in this region due to the sqrt function, leading to more pronounced dithering:

Using 0.0:
![Screenshot from 2020-12-11 17-33-26](https://user-images.githubusercontent.com/21999379/101936162-1d91c800-3bd8-11eb-980c-6eef6f6356f5.png)
Using 0.01:
![Screenshot from 2020-12-11 17-32-50](https://user-images.githubusercontent.com/21999379/101936184-22567c00-3bd8-11eb-9f2d-f80e95600f58.png)

[normal_map_manual_test2.zip](https://github.com/godotengine/godot/files/5680678/normal_map_manual_test2.zip)

**Comment by fuzzypurplepixel** [comment]
`normal.z = sqrt(max (0.03, 1.0 - dot(normal.xy, normal.xy)));`
removes visible noise clumps. I have not been coding many shaders, is it possible to print the values of normal.z?

**Comment by lawnjelly** [comment]
> `normal.z = sqrt(max (0.03, 1.0 - dot(normal.xy, normal.xy)));`
> removes visible noise clumps. I have not been coding many shaders, is it possible to print the values of normal.z?

It's not easy getting data out of a shader. Often all you can do is write data as colour, then screenshot and analyse the colours (while taking into account any colour changes due to tone mapping etc). Perhaps there is some debugger that would help with this with e.g. a software rasterizer but I'm not aware of one.

You can duplicate the shader on the CPU, but it is hard to emulate the effects of precision (I kept meaning to try and have a go adding this to swiftshader or ask them to add it).

**Comment by lawnjelly** [comment]
In your example project to minimize problems you would select the import tab for the normal map and change compress mode to lossless. Texture compression (in addition to the fixed 2 channel compression) makes the artifacts worse. I'm not sure where the sprite 3d normal maps are converted, ah yes, here:

```
#if defined(ENABLE_NORMALMAP)

	normalmap.xy = normalmap.xy * 2.0 - 1.0;
	normalmap.z = sqrt(max(0.01, 1.0 - dot(normalmap.xy, normalmap.xy))); //always ignore Z, as it can be RG packed, Z may be pos/neg, etc.

	normal = normalize(mix(normal, tangent * normalmap.x + binormal * normalmap.y + normal * normalmap.z, normaldepth));

#endif
```
I just tried applying the bodge here in the scene shader, and what do you know, it also clears the artifacts. I'll take a screenshot.

**Comment by fuzzypurplepixel** [comment]
I was able to build godot 3.2 and it seems like 2D is working. Are you still waiting to push the 3D fix

**Comment by lawnjelly** [comment]
> Are you still waiting to push the 3D fix

Yes, I have been away from home. It's not just making a PR, it's discussing it with clayjohn and reduz and checking over the compression code to check there isn't a bug there, or a more appropriate way to address this. In general we don't like to make bug fixes until we are sure it is the right fix. Better to be patient and get the right fix than to rush something through and find it breaks something else.

**Comment by clayjohn** [comment]
I wonder if using a more correct encoding/decoding method would be enough. see https://aras-p.info/texts/CompactNormalStorage.html

**Comment by clayjohn** [comment]
I took another look at this as I was porting the fix to Godot 4.0. The source normalmap is not normalized. So at certain points the length of the normal far exceeds unit length. To save on bandwidth, Godot ignores the z channel and instead reconstructs it from the x and y channels. I'm not sure that this optimization makes sense as for many drivers a texture read will read all 4 channels from the source texture anyway. In 3D, we optimize the normal map to an RG texture, so we know we aren't wasting bandwidth (we also validate the normals at convert time). For 2D we just use the source normalmap and trust that it is proper - thus leading to issues like this one. 

As a workaround for users facing this issue you can insert the following into your fragment shader:
```
vec3 NORMAL = texture(NORMAL_TEXTURE, UV).xyz
NORMAL.xy = NORMAL.xy * 2.0 - 1.0;
```
This will read your source normals directly and should result in nicer looking results.
